tensorrt_llm/auto_parallel/auto_parallel.py

```diff
@@ -0,0 +1,263 @@
+import gc
+import os
+from concurrent.futures import ThreadPoolExecutor
+from pathlib import Path
+
+import tensorrt as trt
+import torch
+from filelock import FileLock
+
+from tensorrt_llm.functional import DimRange, Tensor
+from tensorrt_llm.logger import logger
+from tensorrt_llm.network import Network, net_guard
+
+from .config import AutoParallelConfig
+from .device_mesh import LogicalDeviceMesh, PhysicalDeviceMesh
+from .node_graph import NodeGraph
+from .parallelization import ParallelConfig, parallelize
+from .pipeline_graph import PipelineGraph
+from .simplifier import GraphConfig, Simplifier, StageType
+from .utils import current_flags
+
+
+def to_network(graph: PipelineGraph, network: Network):
+    logger.debug("Converting graph to network")
+    trt_network = graph.as_trt()
+    trt_network.name = network.trt_network.name
+    new_network = Network()
+    new_network._init(trt_network)
+    new_network._dtype = network._dtype
+    new_network._plugin_config = network._plugin_config
+    new_network._unfilled_weights = graph._unfilled_weights
+    new_network._auto_parallel_config = graph._auto_parallel_config
+    with net_guard(network):
+        for i in range(trt_network.num_inputs):
+            input = trt_network.get_input(i)
+            tensor = Tensor(is_network_input=False)
+            if input.name in network._inputs:
+                profiles = network._inputs[input.name].profiles
+            elif len(network._inputs) == 0:
+                profiles = []
+            else:
+                shape = input.shape
+                num_profiles = len(list(network._inputs.values())[0].profiles)
+                profile = DimRange(shape, [None] * len(shape))
+                profiles = [profile] * num_profiles
+            tensor.profiles = profiles
+            tensor.trt_tensor = input
+            new_network._inputs[input.name] = tensor
+    return new_network
+
+
+def find_solution(
```

Comment:
Hi, previously, almost 1 year ago, we also wrote a python/cpp separately to find pipeline partitions upon ONNX IR and cpp IR (where we can have more accurate memory estimation).

Since Tensor-LLM now uses its own memory-pool to mange memory allocation.

The method we use (I also wrote a solver with dyanmical programming) is multi-level graph based.

The problem is performaning solver upon ONNX IR graph is that you cannot always garantee the same topological order in frontend and backend. I guess you will have the same problem.

Obviously one start with 2-device partition, the answer is naive. (that was initial reason why I used dynamic programming), then extend the solution to 4, 8 or many devices.

Iterating the nodes is time consuming, so we have to do it with cpp backend with heuristic search.


I wonder how did you managed to run into these problems ? and what is the compile time you have , for example 1000 nodes graph ?

Reply:
@Shixiaowei02 @kaiyux